### PR TITLE
Add Go template capability for creating functions

### PR DESCRIPTION
**Pull request by nicholasjackson**
## Description
Added capability to create functions in go, a new template has been added and the cli has been modified to allow an additional language 'Go'.

## Motivation and Context
This update adds the capability to create new functions using the Go programming language using the fans-cli.

https://github.com/openfaas/faas-cli/issues/86
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
Unit tests have been run and a manual test has been performed creating a Go function and deploying it to OpenFaaS.  The function was then load tested on the running OpenFaaS instance.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: template/go/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
+FROM golang:1.8.3-alpine
+
+RUN apk --no-cache add curl \ 
+    && echo "Pulling watchdog binary from Github." \
+    && curl -sSL https://github.com/openfaas/faas/releases/download/0.6.5/fwatchdog > /usr/bin/fwatchdog \
+    && chmod +x /usr/bin/fwatchdog \
+    && apk del curl --no-cache
+
```
Comment (austinfrey) [inline template/go/Dockerfile]: one quick suggestion, some other pull requests that are pending are opting to run the functions as non-root users. some examples here by @ericstoekl https://github.com/openfaas/faas-cli/pull/91.

**Comment by alexellis** [comment]
Good point Austin. Maybe we can take that in a follow-up PR?

File: template/go/Dockerfile (hunk 1 above)
Comment (nicholasjackson) [reply template/go/Dockerfile]: Ha, I should know that I was one of the guilty parties behind caninothazroot.

Not even following my own advice

File: template/go/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+COPY . .
+
+# Run a gofmt and exclude all vendored code.
+RUN test -z "$(gofmt -l $(find . -type f -name '*.go' -not -path "./vendor/*"))" || { echo "Run \"gofmt -s -w\" on your Golang code"; exit 1; }
+
+RUN CGO_ENABLED=0 GOOS=linux \
+    go build --ldflags "-s -w" -a -installsuffix cgo -o handler . && \
+    go test ./... -cover
+
+FROM alpine:latest
```
Comment (alexellis) [inline template/go/Dockerfile]: Please pin to 3.6

**Pull request merged**